R/estimate_delay.R

```diff
@@ -0,0 +1,80 @@
+#' Estimate a delay distribution from a reporting triangle
+#' @description
+#' Provides an estimate of the reporting delay as a function
+#' of the delay, based on the reporting triangle and the specified maximum
+#' delay and number of reference date observations to be used in the estimation.
+#' This point estimate of the delay is computed empirically, using an
+#' iterative algorithm starting from the most recent observations. It was
+#' modified from the code originally developed by the Karlsruhe Institute
+#' of Technology RESPINOW German Hospitalization Nowcasting Hub,
+#' Modified from: https://github.com/KITmetricslab/RESPINOW-Hub/blob/7cce3ae2728116e8c8cc0e4ab29074462c24650e/code/baseline/functions.R#L55 #nolint
+#' @param triangle dataframe of the reporting triangle, with one column for the
+#' reference date and the remaining columns specifying the delay (units not
+#' specified). Values indicate the number of new observations assigned to
+#' the reference date at the specified delay.
+#' @param max_delay integer indicating the maximum delay to estimate, in units
+#' of the delay. The default is to use the whole reporting triangle,
+#'  `ncol(triangle) -1`.
+#' @param n_history integer indicating the number of reference dates to be
+#' used in the estimate of the reporting delay, always starting from the most
+#' recent reporting delay. The default is to use the whole reporting triangle,
+#' so `nrow(triangle)-1`
+#' @returns delay_df dataframe of length `max_delay` with columns `delay`
+#' and `pmf`, indicating the point estimate of the empirical probability
+#' mass on each delay
+#' @export
+#' @importFrom dplyr filter
+#' @examples
+#' library(epinowcast)
+#' nat_germany_hosp <-
+#'   germany_covid19_hosp[location == "DE"][age_group == "00+"]
+#' nat_germany_hosp <- enw_filter_report_dates(
+#'   nat_germany_hosp,
+#'   latest_date = "2021-10-01"
+#' )
+#' pobs <- enw_preprocess_data(nat_germany_hosp, max_delay = 21)
+#' triangle_raw <- pobs$reporting_triangle[[1]]
+#' delay_df <- estimate_delay(triangle_raw[, -1],
+#'   max_delay = 20,
+#'   n_history = 30
+#' )
+estimate_delay <- function(triangle,
+                           max_delay = ncol(triangle) - 2,
+                           n_history = nrow(triangle)) {
+  # Check that the input reporting triangle is formatted properly.
+  validate_triangle(triangle)
+
+  # Filter the triangle down to nrow = n_history + 1, ncol = max_delay
+  max_date <- max(triangle$reference_date, na.rm = TRUE) - n_history
+  trunc_triangle <- preprocess_reporting_triangle(triangle, max_delay) |>
+    dplyr::filter(reference_date >= max_date) |>
```

Comment:
use .data$

Reply:
Refactored this per our f2f discussion, removing the use of `dplyr` internally and expecting the lowest level to start (just the reporting triangle matrix)